File: SoloLearn Codes/Digits of PI/digits_of_pi.swift
import Foundation

var arr = (0...3501).map { x in 2000 }
var piStr = String()
var carry = 0
for i in stride(from: 3500, to: 0, by: -14) {
    var sum = 0
    for j in stride(from: i, to: 0, by: -1) {
        sum = sum * j + 10000 * arr[j]
        arr[j] = sum % (j * 2 - 1)
        sum /= j * 2 - 1
    }
    piStr += String(format: "%04d", carry + sum / 10000)
    carry = sum % 10000
}
print(Array(piStr)[Int(readLine()!)!])

File: SoloLearn Codes/Digits of PI/DigitsOfPi.cs
using System;
using System.Linq;
using System.Text;

namespace SoloLearn
{
    class Program
    {
        static void Main(string[] args)
        {
            int[] arr = Enumerable.Repeat(2000, 3500 + 1).ToArray();
            StringBuilder sb = new StringBuilder();
            int carry = 0;
            for (int i = 3500; i > 0; i -= 14)
            {
                int sum = 0;
                for(int j = i; j > 0; j--)
                {
                    sum = sum * j + 10000 * arr[j];
                    arr[j] = sum % (j * 2 - 1);
                    sum /= j * 2 - 1;
                }
                sb.Append($"{carry + sum / 10000:D4}");
                carry = sum % 10000;
            }
            Console.WriteLine(sb.ToString()[Convert.ToInt32(Console.ReadLine())]);
        }
    }
}